---
title: "A small real data example"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{real-data-example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 300px;
}
```

The overall process should be <b>very straight forward</b>: 

1. We first run Relate starting from a phased VCF file
2. Then we run Twigstats on these inferred trees, using a single R function called ```f2_blocks_from_Relate```. 
3. This object is then fed into admixtools2 to compute <b>any</b> f-statistic.

We will illustrate how to run Twigstats on a small example of imputed ancient genomes.
We will use chromosome 1 of 25 publicly available low coverage shotgun genomes from Anglo-Saxon contexts, British Iron/Roman Age, Irish Bronze Age, and the Scandinavian Early Iron Age (Cassidy et al, PNAS 2016; Martiniano et al, Nature Communications 2016; Anastasiadou et al, Communications Biology 2023; Schiffels et al Nature Communications 2016; Gretzinger et al Nature 2022; Rodriguez-Varela et al Cell 2023). 

To look up where preprepared input files are stored on your computer, please clone the Twigstats github repository and execute the following command:
```
git checkout documentation
print(system.file("example/", package = "twigstats"))
```

In this example, we have already executed the steps detailed under "Preparing the data", so you can skip ahead to "Running Relate".

## Preparing the data

Here, we will follow the standard pipeline for imputing ancient shotgun genomes. These steps have already been executed and you can skip ahead to "Running Relate" if you like. 

#### Imputation

We imputed the ancient genomes with GLIMPSE (Rubinacci et al, Nature Genetics, 2021; https://odelaneau.github.io/GLIMPSE/) and using the 1000 Genomes Project reference panel. We then merged these imputed genomes with a subset of the 1000 Genomes Project dataset. You can see the genomes we will use here in the box below:

```{r, style="max-height: 200px;"}
popl <- read.table(system.file("example/1000GP_Phase3_sub_aDNA.poplabels", package = "twigstats"), header = T)
print(popl)
```

Given this merged VCF, we then remove SNPs with missing data, and filter the data by imputation quality (e.g. requiring a minimum imputation posterior or minimum [INFO score](https://static-content.springer.com/esm/art%3A10.1038%2Fnrg2796/MediaObjects/41576_2010_BFnrg2796_MOESM3_ESM.pdf)).

Removing all SNPs where any genotype has an imputation posterior <80% is achieved by running something like the following command
```  
bcftools view --max-alleles 2 --exclude-types indels 1000G_aDNA_chr1.vcf.gz | 
bcftools +setGT -- -t q -n . -i'SMPL_MAX(FORMAT/GP)<=0.8' | 
bcftools view -i 'F_MISSING==0.0' -O z -o 1000G_aDNA_filtered_chr1.vcf.gz
```
This usually works well with a small number of samples. 

<b>For larger data sets</b>, this is likely too restrictive. Please note that Relate cannot accept missing data as input, so if you would like to e.g. retain only sites where <2% of genotypes have a posterior <80%, then you will need to first generate a list of BP positions of SNPs to keep:
```
bcftools view --max-alleles 2 --exclude-types indels 1000G_aDNA_chr1.vcf.gz | 
bcftools +setGT -- -t q -n . -i'SMPL_MAX(FORMAT/GP)<=0.8' | 
bcftools view -i 'F_MISSING<=0.02' | 
bcftools query -f "%CHROM\t%POS\n" > BP_chr1.txt
```
We can then subset the original imputed vcf using bcftools:
```
bcftools view -R BP_chr1.txt 1000G_aDNA_chr1.vcf.gz -O z -o 1000G_aDNA_filtered_chr1.vcf.gz
```

Other filtering approaches are also possible, e.g., we can use bcftools to annotate each SNP with an INFO score (see [impute-info](https://samtools.github.io/bcftools/howtos/plugins.html) plugin of bcftools) and require a minimum INFO score for each SNP. This score measures how well a SNP is imputed using surrounding haplotype information. 

#### Genome mask and ancestral genome

You can download the 1000 Genomes Project accessibility masks from this URL for [hg37](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/) and [hg38](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/working/20160622_genome_mask_GRCh38/).

The ancestral genome is available here for [hg37](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/supporting/ancestral_alignments/) and [hg38](https://ftp.ensembl.org/pub/release-86/fasta/ancestral_alleles/).

These files, and a recombination map for hg37 and hg38, can also be downloaded from [here](https://www.dropbox.com/scl/fi/ocvr7m00f4jhfcytew4fy/Relate_input_files.tar?rlkey=j1kuqbo2hgvgfwmudwao0v0wv&dl=0).

#### Subsetting to transversions (optional)

We may also opt to only analyse transversions, which we can extract using the following command
```
bcftools view -i 'TYPE="snp" && ( (REF!="A" || ALT!="G") && (REF!="G" || ALT!="A") && (REF!="C" || ALT!="T") && (REF!="T" || ALT!="C"))' 1000G_aDNA_filtered_chr1.vcf.gz -O z -o 1000G_aDNA_transv_chr1.vcf.gz
```

#### Conversion to Relate input format

Once we have a VCF containing SNPs that we are happy with, we can now use a script provided by the Relate package to convert this into the Relate input format. For details please consult the [Relate documentation](https://myersgroup.github.io/relate/):
```
#convert from vcf to haps/sample
~/Documents/genomics/software/relate/bin/RelateFileFormats \
    --mode ConvertFromVcf \
  -i 1000G_aDNA_transv_chr1 \
	--haps 1000G_aDNA_transv_chr1.haps \
	--sample 1000G_aDNA_transv_chr1.sample

gzip -f 1000G_aDNA_transv_chr1.haps
gzip -f 1000G_aDNA_transv_chr1.sample

#apply a genomic mask and polarise alleles according to ancestral/derived state
~/Documents/genomics/software/relate/scripts/PrepareInputFiles/PrepareInputFiles.sh \
    --haps 1000G_aDNA_transv_chr1.haps.gz \
	--sample 1000G_aDNA_transv_chr1.sample.gz \
	--ancestor human_ancestor_chr1.fa.gz \
	-o 1000G_aDNA_mask_transv_chr1 \
	--remove_ids remove.txt \
	--mask ./StrictMask_chr1.fa.gz
```
Here, we used a [human ancestral genome](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/supporting/ancestral_alignments/) and [mappability mask](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/). These files are avaliable for GRCh38 as well: [human ancestral genome](https://ftp.ensembl.org/pub/release-86/fasta/ancestral_alleles/) and [mappability mask](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/working/20160622_genome_mask_GRCh38/).

## Running Relate

Once we have prepared the data, we can now run Relate (Speidel et al, Nature Genetics 2019; MBE 2021). Please consult the [Relate documentation](https://myersgroup.github.io/relate/) for details on how to run Relate. Here, we will run Relate using a mutation rate of 4e-9, corresponding approximately to the average transversion mutation rate in humans, and we will use pre-inferred coalescence rates stored in ```1000G_auto.coal```.

```{bash}
path="/opt/homebrew/lib/R/4.3/site-library/twigstats/example/"

Relate --mode All \
    --haps ${path}/1000G_aDNA_mask_transv_chr1.haps.gz \
	--sample ${path}/1000G_aDNA_mask_transv_chr1.sample.gz \
	--dist ${path}/1000G_aDNA_mask_transv_chr1.dist.gz \
	--map ${path}/genetic_map_combined_b37_chr1.txt \
	-o 1000G_aDNA_mask_transv_chr1 \
	-m 4e-9 \
	--coal ${path}/1000G_auto.coal \
	--sample_ages ${path}/sample_ages.txt \
	--memory 5 \
	--seed 1

gzip -f 1000G_aDNA_mask_transv_chr1.anc
gzip -f 1000G_aDNA_mask_transv_chr1.mut
```

## Running Twigstats

Finally, once we have inferred Relate trees, we can now run Twigstats. 
We will aim to infer an admixture in Anglo-Saxon individuals between the preceding British Iron/Roman Age and Early Iron Age Scandinavia.

To illustrate a power gain by using Twigstats, we will run Twigstats in two different ways: First, we will compute regular f2-statistics and second, we will compute f2-statistics on only recent coalescences.

We first load the Twigstats package and assign filenames to variables.
```{r}
library(twigstats)
file_anc  <- "1000G_aDNA_mask_transv_chr1.anc.gz"
file_mut  <- "1000G_aDNA_mask_transv_chr1.mut.gz"
poplabels <- system.file("example/1000GP_Phase3_sub_aDNA.poplabels", package = "twigstats")
file_map  <- system.file("example/genetic_map_combined_b37_chr1.txt", package = "twigstats")
```

Next, we compute regular f2 statistics without a time cutoff, and then infer the admixture proportion using an f4-ratio statistic:
```{r}
#Calculate f2s between all pairs of populations
f2_blocks <- f2_blocks_from_Relate(file_anc, file_mut, poplabels, file_map, use_muts = T)
f4_ratio(f2_blocks, popX="England_Saxon.SG", popI="Ireland_BA.SG", pop1="Britain.IronRoman.SG", pop2="Scandinavian_Peninsula_EIA(I).SG", popO="YRI")
```
As we can see, the inferred admixture proportion is not between 0 and 1 and the standard errors are extremely large, giving us no meaningful inference.

We now compute f2 statistics on lineages of the past 500 generations using Twigstats:
```{r}
#Use a cutoff of 500 generations
f2_blocks <- f2_blocks_from_Relate(file_anc, file_mut, poplabels, file_map, t = 500)
f4_ratio(f2_blocks, popX="England_Saxon.SG", popI="Ireland_BA.SG", pop1="Britain.IronRoman.SG", pop2="Scandinavian_Peninsula_EIA(I).SG", popO="YRI")
```
The standard errors are much reduced and we get a meaningful <b>admixture proportion of approximately 38% Scandinavian EIA ancestry</b>! Note that the standard error is still relatively large (though excludes 0 and 1 as feasible proportions). The obvious way to improve power is to use all other chromosomes; here we only used chromosome 1. Additionally, we could build Relate trees with more samples to get more accurate constrains on dates.

## Multiple chromosomes

If you have multiple chromosomes, you can use the argument ```chr``` in Twigstats functions, which takes an R array with chromosome names as input. For ```file_anc```, ```file_mut```, and ```file_map```, you will only need to specify the filename prefix, such that e.g., the anc files are named ```paste0(file_anc,"_chr1.anc.gz")``` etc.

E.g., the above Twigstats runs are equivalent to
```{r}
file_anc  <- "1000G_aDNA_mask_transv" #filename prefix only
file_mut  <- "1000G_aDNA_mask_transv" #filename prefix only

path      <- paste0(system.file("example/", package = "twigstats"),"/") #just necessary to get the path to the poplabels file
poplabels <- paste0(path,"1000GP_Phase3_sub_aDNA.poplabels")
file_map  <- paste0(path,"genetic_map_combined_b37") #filename prefix only
chrs <- 1:1

#Calculate f2s between all pairs of populations
#We specify the argument chr, which takes an R array as input
f2_blocks <- f2_blocks_from_Relate(file_anc, file_mut, poplabels, file_map, chr = chrs, t = 500)

#Once we have f2_blocks we can compute f-statistics as before
f4_ratio(f2_blocks, popX="England_Saxon.SG", popI="Ireland_BA.SG", pop1="Britain.IronRoman.SG", pop2="Scandinavian_Peninsula_EIA(I).SG", popO="YRI")

```


