<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic usage • twigstats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic usage">
<meta property="og:description" content="twigstats">
<meta property="og:image" content="https://leospeidel.github.io/twigstats/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-P8RP1CGQ1W"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-P8RP1CGQ1W');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">twigstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-usage.html">Basic usage</a>
    </li>
    <li>
      <a href="../articles/real-data-example.html">A small real data example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/leospeidel/twigstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic usage</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/leospeidel/twigstats/blob/HEAD/vignettes/basic-usage.Rmd" class="external-link"><code>vignettes/basic-usage.Rmd</code></a></small>
      <div class="hidden name"><code>basic-usage.Rmd</code></div>

    </div>

    
    
<p>The main use-case of Twigstats is to compute <b>f2-statistics for
pairs of populations</b>. The output can then directly be fed into
functions of the admixtools2 R package (<a href="https://uqrmaie1.github.io/admixtools/index.html" class="external-link uri">https://uqrmaie1.github.io/admixtools/index.html</a>) to
<b>compute any other f-statistic</b>.<br></p>
<p>Each function implemented in the Twigstats package is described on
this page: <a href="https://leospeidel.com/twigstats_old/reference" class="external-link">leospeidel.com/twigstats/reference</a>.</p>
<p>Here, we will demonstrate key features using an example stored under
<code>inst/sim/</code>, which contains simulated data for five
populations as described in our paper (Figure 1).</p>
<center>
<img src="../../../../../../../../UCL%20Dropbox/Leo%20Speidel/Mac%20(2)/Documents/genomics/software/twigstats/articles/images/f4ratio.png" id="id" class="class" width="300" height="300">
</center>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/leospeidel/twigstats" class="external-link">twigstats</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
<div class="section level2">
<h2 id="computing-f2-statistics-from-relate-genealogies">Computing f2-statistics from Relate genealogies<a class="anchor" aria-label="anchor" href="#computing-f2-statistics-from-relate-genealogies"></a>
</h2>
<p>The first use-case is to compute f2-statistics on Relate genealogies
(<a href="https://myersgroup.github.io/relate/" class="external-link uri">https://myersgroup.github.io/relate/</a>).<br></p>
<p>The central function is called <code><a href="../reference/f2_blocks_from_Relate.html">f2_blocks_from_Relate()</a></code>
which takes Relate output files as input.<br> Please type
<code><a href="../reference/f2_blocks_from_Relate.html">?f2_blocks_from_Relate</a></code> for more details:<br></p>
<p>We will illustrate a basic Twigstats run using example files that are
included in the Twigstats package. In this example, population PX is
admixed between populations P2 and P3, with a mixture proportion of 20%
and 80%, respectively.</p>
<p>These lines assign filenames to variables <code>file_anc</code>,
<code>file_mut</code>, <code>poplabels</code>, and
<code>file_map</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#see https://myersgroup.github.io/relate/getting_started.html#Output for file formats</span></span>
<span><span class="va">file_anc</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/msprime_ad0.8_split250_1_chr1.anc.gz"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span></span>
<span><span class="va">file_mut</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/msprime_ad0.8_split250_1_chr1.mut.gz"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span></span>
<span><span class="co">#see https://myersgroup.github.io/relate/input_data.html for file formats</span></span>
<span><span class="va">poplabels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/msprime_ad0.8_split250_1.poplabels"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span></span>
<span><span class="va">file_map</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/genetic_map_combined_b37_chr1.txt"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span> <span class="co">#recombination map (three column format)</span></span></code></pre></div>
<p>We will first calculate f2 statistics as normal without a
cutoff.<br> We will then plug these f2 statistics into an f4-ratio
statistic which computes an estimate of the admixture proportion.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Calculate regular f2s between all pairs of populations</span></span>
<span><span class="va">f2_blocks1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_Relate.html">f2_blocks_from_Relate</a></span><span class="op">(</span><span class="va">file_anc</span>, <span class="va">file_mut</span>, <span class="va">poplabels</span>, <span class="va">file_map</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks1</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO popI pop1 pop2 popX       val        se</span></span>
<span><span class="co">#&gt; 1   P4   P1   P2   P3   PX 0.9974217 0.3076228</span></span></code></pre></div>
<p>As we can see, the inferred admixture proportion is not close to the
true value of 80% and the standard errors are large. This shows that we
are underpowered to infer the admixture proportions using this f4-ratio
statistic.</p>
<p>Now we will specify a twigstats cutoff of 500 generations and plug
these into an f4-ratio statistic:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Now calculate f2s using a twigstats cutoff of 500 generations. </span></span>
<span><span class="co">#This should give us a big power boost.</span></span>
<span><span class="va">f2_blocks2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_Relate.html">f2_blocks_from_Relate</a></span><span class="op">(</span><span class="va">file_anc</span>, <span class="va">file_mut</span>, <span class="va">poplabels</span>, <span class="va">file_map</span>, t <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks2</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO popI pop1 pop2 popX       val         se</span></span>
<span><span class="co">#&gt; 1   P4   P1   P2   P3   PX 0.8110266 0.02085804</span></span></code></pre></div>
<p>The standard errors are much reduced!</p>
<ul>
<li>Use argument <b>use_muts</b> to compute f2-statistics on the
(age-ascertained) mutations, instead of branch lengths of Relate
genealogies.</li>
<li>Use argument <b>blgsize</b> to change jackknife block size</li>
<li>Set argument <b>transitions</b> to False to exclude transitions</li>
</ul>
<div class="section level3">
<h3 id="using-admixtools2">Using admixtools2<a class="anchor" aria-label="anchor" href="#using-admixtools2"></a>
</h3>
<p>We can now directly feed this into admixtools2 functions:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/uqrmaie1/admixtools" class="external-link">admixtools</a></span><span class="op">)</span></span>
<span><span class="co">#no twigstats ascertainment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/admixtools/man/qpdstat.html" class="external-link">f4</a></span><span class="op">(</span><span class="va">f2_blocks1</span>, <span class="st">"P4"</span>, <span class="st">"P3"</span>, <span class="st">"P2"</span>, <span class="st">"PX"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   pop1  pop2  pop3  pop4      est       se     z         p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4    P3    P2    PX    0.001<span style="text-decoration: underline;">27</span> 0.000<span style="text-decoration: underline;">314</span>  4.04 0.000<span style="text-decoration: underline;">054</span>3</span></span>
<span><span class="co">#with twigstats ascertainment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/admixtools/man/qpdstat.html" class="external-link">f4</a></span><span class="op">(</span><span class="va">f2_blocks2</span>, <span class="st">"P4"</span>, <span class="st">"P3"</span>, <span class="st">"P2"</span>, <span class="st">"PX"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   pop1  pop2  pop3  pop4       est        se     z         p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4    P3    P2    PX    0.000<span style="text-decoration: underline;">635</span> 0.000<span style="text-decoration: underline;">025</span>3  25.1 1.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-138</span></span></span></code></pre></div>
<p>If PX exclusively clustered with P2 and did not carry gene flow from
P3, we expect this f4 statistic to not be significantly deviating from
0. We observe a highly significant attraction of P3 to PX, especially
when ascertaining coalescences of the last 500 generations using
Twigstats, consistent with PX being a mixture of P2 and P3.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="computing-f2-statistics-on-age-ascertained-snps">Computing f2-statistics on age-ascertained SNPs<a class="anchor" aria-label="anchor" href="#computing-f2-statistics-on-age-ascertained-snps"></a>
</h2>
<p>Sometimes, we are prefer not to graft samples into genealogies
(e.g. due to low sequencing coverage, or if we don’t trust the
imputation). <br> Therefore, the second use-case of Twigstats is to
compute f2-statistics on age ascertained mutations.<br></p>
<p>The central function is called
<code><a href="../reference/f2_blocks_from_RelateAges.html">f2_blocks_from_RelateAges()</a></code> which takes plink (bed/bim/fam)
files and mutation ages (.mut) as input.<br> Please type
<code><a href="../reference/f2_blocks_from_RelateAges.html">?f2_blocks_from_RelateAges</a></code> for more details:<br></p>
<p>We will begin with assigning filenames to the following variables. In
our case, all files are stored under <code>path</code> and have prefix
<code>pref</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span>,<span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">pref</span> <span class="op">&lt;-</span> <span class="st">"msprime_ad0.8_split250_1"</span></span>
<span><span class="va">file_plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix</span></span>
<span><span class="va">file_mut</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix (here same name as plink file but can be different)</span></span></code></pre></div>
<p>Next, we compute f2-statistics as normal on the plink format and plug
these into an f4-ratio statistic to estimate the admixture
proportion.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Compute regular f2 statistics between all pairs of populations. You can use pops to only calculate f2s between specified populations.</span></span>
<span><span class="va">f2_blocks1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_RelateAges.html">f2_blocks_from_RelateAges</a></span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks1</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO popI pop1 pop2 popX      val        se</span></span>
<span><span class="co">#&gt; 1   P4   P1   P2   P3   PX 1.082681 0.2931867</span></span></code></pre></div>
<p>As before, we can see that the standard error of the inferred
dmixture proportion is very large. This shows we are lacking power using
the standard f-statistics approach.</p>
<p>We can now ascertain mutations by age. Here we choose a cutoff of 500
generations:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Use a cutoff of 500 generations</span></span>
<span><span class="va">f2_blocks2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_RelateAges.html">f2_blocks_from_RelateAges</a></span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span>, t <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks2</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO popI pop1 pop2 popX      val         se</span></span>
<span><span class="co">#&gt; 1   P4   P1   P2   P3   PX 0.810077 0.03097378</span></span></code></pre></div>
<p>The standard errors are much reduced! So while we are unable to get a
reliable estimate without twigstats ascertainment, we estimate PX to
carry approximately 80% P2 and 20% P3 ancestry with twigstats.</p>
<ul>
<li>Use argument <b>pops</b> to analyse a subset of populations and
<b>fam</b> to change population assignment of individuals</li>
<li>Use argument <b>blgsize</b> to change jackknife block size</li>
<li>Set argument <b>transitions</b> to False to exclude transitions</li>
<li>Use file_mut = “./1240k/1240k” to calculate f2 statistics at 1240k
sites</li>
<li>Use file_mut = “./1000G_ages/1000GP_Phase3_mask_prene_nosingle_chr1”
to calculate f2 statistics at 1240k sites</li>
</ul>
<div class="section level3">
<h3 id="using-admixtools2-1">Using admixtools2<a class="anchor" aria-label="anchor" href="#using-admixtools2-1"></a>
</h3>
<p>We can now directly feed this into admixtools2 functions:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/uqrmaie1/admixtools" class="external-link">admixtools</a></span><span class="op">)</span></span>
<span><span class="co">#no twigstats ascertainment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/admixtools/man/qpdstat.html" class="external-link">f4</a></span><span class="op">(</span><span class="va">f2_blocks1</span>, <span class="st">"P4"</span>, <span class="st">"P3"</span>, <span class="st">"P2"</span>, <span class="st">"PX"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   pop1  pop2  pop3  pop4      est       se     z        p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4    P3    P2    PX    0.001<span style="text-decoration: underline;">16</span> 0.000<span style="text-decoration: underline;">305</span>  3.81 0.000<span style="text-decoration: underline;">139</span></span></span>
<span><span class="co">#with twigstats ascertainment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/admixtools/man/qpdstat.html" class="external-link">f4</a></span><span class="op">(</span><span class="va">f2_blocks2</span>, <span class="st">"P4"</span>, <span class="st">"P3"</span>, <span class="st">"P2"</span>, <span class="st">"PX"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   pop1  pop2  pop3  pop4       est        se     z         p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> P4    P3    P2    PX    0.000<span style="text-decoration: underline;">642</span> 0.000<span style="text-decoration: underline;">028</span>1  22.9 9.35<span style="color: #949494;">e</span><span style="color: #BB0000;">-116</span></span></span></code></pre></div>
<p>As before, we find a highly significant attraction of P3 to PX using
Twigstats, consistent with PX being a mixture of P2 and P3.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Leo Speidel, Pontus Skoglund.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
