<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A small real data example • twigstats</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A small real data example">
<meta property="og:description" content="twigstats">
<meta property="og:image" content="https://leospeidel.github.io/twigstats/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-P8RP1CGQ1W"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-P8RP1CGQ1W');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">twigstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-usage.html">Basic usage</a>
    </li>
    <li>
      <a href="../articles/real-data-example.html">A small real data example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/leospeidel/twigstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A small real data example</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/leospeidel/twigstats/blob/HEAD/vignettes/real-data-example.Rmd" class="external-link"><code>vignettes/real-data-example.Rmd</code></a></small>
      <div class="hidden name"><code>real-data-example.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre {
  max-height: 300px;
  overflow-y: auto;
}

pre[class] {
  max-height: 300px;
}
</style>
<p>We will illustrate how to run Twigstats on a small example of imputed
ancient genomes. We will use chromosome 1 of 25 publicly available low
coverage shotgun genomes from Anglo-Saxon contexts, British Iron/Roman
Age, Irish Bronze Age, and the Scandinavian Early Iron Age (Cassidy et
al, PNAS 2016; Martiniano et al, Nature Communications 2016;
Anastasiadou et al, Communications Biology 2023; Schiffels et al Nature
Communications 2016; Gretzinger et al Nature 2022; Rodriguez-Varela et
al Cell 2023).</p>
<p>To look up where preprepared input files are stored on your computer,
please execute the following command once Twigstats is installed:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"example/"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>In this example, we have already executed the steps detailed under
“Preparing the data”, so you can skip ahead to “Running Relate”.</p>
<div class="section level2">
<h2 id="preparing-the-data">Preparing the data<a class="anchor" aria-label="anchor" href="#preparing-the-data"></a>
</h2>
<p>Here, we will follow the standard pipeline for imputing ancient
shotgun genomes. These steps have already been executed and you can skip
ahead to “Running Relate” if you like.</p>
<div class="section level4">
<h4 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a>
</h4>
<p>We imputed the ancient genomes with GLIMPSE (Rubinacci et al, Nature
Genetics, 2021; <a href="https://odelaneau.github.io/GLIMPSE/" class="external-link uri">https://odelaneau.github.io/GLIMPSE/</a>) and using the 1000
Genomes Project reference panel. We then merged these imputed genomes
with a subset of the 1000 Genomes Project dataset. You can see the
genomes we will use here in the box below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">popl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"example/1000GP_Phase3_sub_aDNA.poplabels"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">popl</span><span class="op">)</span></span>
<span><span class="co">#&gt;         ID                              POP   GROUP    SEX</span></span>
<span><span class="co">#&gt; 1  NA12249                              CEU     EUR female</span></span>
<span><span class="co">#&gt; 2  NA12830                              CEU     EUR female</span></span>
<span><span class="co">#&gt; 3  NA07048                              CEU     EUR   male</span></span>
<span><span class="co">#&gt; 4  NA12815                              CEU     EUR female</span></span>
<span><span class="co">#&gt; 5  NA06986                              CEU     EUR   male</span></span>
<span><span class="co">#&gt; 6  NA18635                              CHB     EAS   male</span></span>
<span><span class="co">#&gt; 7  NA18609                              CHB     EAS   male</span></span>
<span><span class="co">#&gt; 8  NA18593                              CHB     EAS female</span></span>
<span><span class="co">#&gt; 9  NA18546                              CHB     EAS   male</span></span>
<span><span class="co">#&gt; 10 NA18595                              CHB     EAS female</span></span>
<span><span class="co">#&gt; 11 HG00233                              GBR     EUR female</span></span>
<span><span class="co">#&gt; 12 HG00151                              GBR     EUR   male</span></span>
<span><span class="co">#&gt; 13 HG00118                              GBR     EUR female</span></span>
<span><span class="co">#&gt; 14 HG00160                              GBR     EUR   male</span></span>
<span><span class="co">#&gt; 15 HG00140                              GBR     EUR   male</span></span>
<span><span class="co">#&gt; 16 NA18907                              YRI     AFR female</span></span>
<span><span class="co">#&gt; 17 NA19257                              YRI     AFR female</span></span>
<span><span class="co">#&gt; 18 NA18504                              YRI     AFR   male</span></span>
<span><span class="co">#&gt; 19 NA18498                              YRI     AFR   male</span></span>
<span><span class="co">#&gt; 20 NA19131                              YRI     AFR female</span></span>
<span><span class="co">#&gt; 21   3DT16             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 22   6DT18             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 23   6DT21             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 24   6DT22             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 25   6DT23             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 26  C10427             Britain.IronRoman.SG      UK      U</span></span>
<span><span class="co">#&gt; 27   I0156             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 28   I0160             Britain.IronRoman.SG      UK      M</span></span>
<span><span class="co">#&gt; 29   I0789             Britain.IronRoman.SG      UK      F</span></span>
<span><span class="co">#&gt; 30   M1489             Britain.IronRoman.SG      UK      F</span></span>
<span><span class="co">#&gt; 31  HAD001                 England_Saxon.SG      UK      M</span></span>
<span><span class="co">#&gt; 32  HAD013                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 33   I0161                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 34   I0774                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 35   I0777                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 36  OAI004                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 37  OAI007                 England_Saxon.SG      UK      F</span></span>
<span><span class="co">#&gt; 38   rath1                    Ireland_BA.SG Ireland      M</span></span>
<span><span class="co">#&gt; 39   rath2                    Ireland_BA.SG Ireland      M</span></span>
<span><span class="co">#&gt; 40   rath3                    Ireland_BA.SG Ireland      M</span></span>
<span><span class="co">#&gt; 41  als001 Scandinavian_Peninsula_EIA(I).SG  Sweden      M</span></span>
<span><span class="co">#&gt; 42  lov001 Scandinavian_Peninsula_EIA(I).SG  Sweden      F</span></span>
<span><span class="co">#&gt; 43  snb010 Scandinavian_Peninsula_EIA(I).SG  Sweden      M</span></span>
<span><span class="co">#&gt; 44  snb013 Scandinavian_Peninsula_EIA(I).SG  Sweden      M</span></span>
<span><span class="co">#&gt; 45  snb014 Scandinavian_Peninsula_EIA(I).SG  Sweden      M</span></span></code></pre></div>
<p>Given this merged VCF, we then remove SNPs with missing data, and
filter the data by imputation quality (e.g. requiring a minimum
imputation posterior or minimum <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fnrg2796/MediaObjects/41576_2010_BFnrg2796_MOESM3_ESM.pdf" class="external-link">INFO
score</a>).</p>
<p>Removing all SNPs where any genotype has an imputation posterior
&lt;80% is achieved by running something like the following command</p>
<pre><code>bcftools view --max-alleles 2 --exclude-types indels 1000G_aDNA_chr1.vcf.gz | 
bcftools +setGT -- -t q -n . -i'SMPL_MAX(FORMAT/GP)&lt;=0.8' | 
bcftools view -i 'F_MISSING==0.0' -O z -o 1000G_aDNA_nomiss_chr1.vcf.gz</code></pre>
<p>This usually works well with a small number of samples. For larger
data sets, this is likely too restrictive. Please note that Relate
cannot accept missing data as input, so if you would like to e.g. retain
only sites where &lt;2% of genotypes have a posterior &lt;80%, then you
will need to first generate a list of BP positions of SNPs to keep, and
then subset the original imputed vcf using bcftools. Such a list can be
generated e.g. by</p>
<pre><code>bcftools view --max-alleles 2 --exclude-types indels 1000G_aDNA_chr1.vcf.gz | 
bcftools +setGT -- -t q -n . -i'SMPL_MAX(FORMAT/GP)&lt;=0.8' | 
bcftools view -i 'F_MISSING&lt;=0.02' | 
bcftools query -f "%CHROM %BP\n" &gt; BP_chr1.txt</code></pre>
<p>We could alternatively use bcftools to annotate each SNP with an INFO
score (see <a href="https://samtools.github.io/bcftools/howtos/plugins.html" class="external-link">impute-info</a>
plugin of bcftools) and require a minimum INFO score for each SNP. This
score measures how well a SNP is imputed using surrounding haplotype
information.</p>
</div>
<div class="section level4">
<h4 id="subsetting-to-transversions-optional">Subsetting to transversions (optional)<a class="anchor" aria-label="anchor" href="#subsetting-to-transversions-optional"></a>
</h4>
<p>We may also opt to only analyse transversions, which we can extract
using the following command</p>
<pre><code>bcftools view -i 'TYPE="snp" &amp;&amp; ( (REF!="A" || ALT!="G") &amp;&amp; (REF!="G" || ALT!="A") &amp;&amp; (REF!="C" || ALT!="T") &amp;&amp; (REF!="T" || ALT!="C"))' 1000G_aDNA_nomiss_chr1.vcf.gz -O z -o 1000G_aDNA_transv_chr1.vcf.gz</code></pre>
</div>
<div class="section level4">
<h4 id="conversion-to-relate-input-format">Conversion to Relate input format<a class="anchor" aria-label="anchor" href="#conversion-to-relate-input-format"></a>
</h4>
<p>Once we have a VCF containing SNPs that we are happy with, we can now
use a script provided by the Relate package to convert this into the
Relate input format. For details please consult the <a href="https://myersgroup.github.io/relate/" class="external-link">Relate
documentation</a>:</p>
<pre><code>#convert from vcf to haps/sample
~/Documents/genomics/software/relate/bin/RelateFileFormats \
    --mode ConvertFromVcf \
  -i 1000G_aDNA_transv_chr1 \
    --haps 1000G_aDNA_transv_chr1.haps \
    --sample 1000G_aDNA_transv_chr1.sample

gzip -f 1000G_aDNA_transv_chr1.haps
gzip -f 1000G_aDNA_transv_chr1.sample

#apply a genomic mask and polarise alleles according to ancestral/derived state
~/Documents/genomics/software/relate/scripts/PrepareInputFiles/PrepareInputFiles.sh \
    --haps 1000G_aDNA_transv_chr1.haps.gz \
    --sample 1000G_aDNA_transv_chr1.sample.gz \
    --ancestor human_ancestor_chr1.fa.gz \
    -o 1000G_aDNA_mask_transv_chr1 \
    --remove_ids remove.txt \
    --mask ./StrictMask_chr1.fa.gz</code></pre>
<p>Here, we used a <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/supporting/ancestral_alignments/" class="external-link">human
ancestral genome</a> and <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/accessible_genome_masks/" class="external-link">mappability
mask</a>. These files are avaliable for GRCh38 as well: <a href="https://ftp.ensembl.org/pub/release-86/fasta/ancestral_alleles/" class="external-link">human
ancestral genome</a> and <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/working/20160622_genome_mask_GRCh38/" class="external-link">mappability
mask</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-relate">Running Relate<a class="anchor" aria-label="anchor" href="#running-relate"></a>
</h2>
<p>Once we have prepared the data, we can now run Relate (Speidel et al,
Nature Genetics 2019; MBE 2021). Please consult the <a href="https://myersgroup.github.io/relate/" class="external-link">Relate documentation</a> for
details on how to run Relate. Here, we will run Relate using a mutation
rate of 4e-9, corresponding approximately to the average transversion
mutation rate in humans, and we will use pre-inferred coalescence rates
stored in <code>1000G_auto.coal</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="va">path</span><span class="op">=</span><span class="st">"/opt/homebrew/lib/R/4.3/site-library/twigstats/example/"</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="ex">Relate</span> <span class="at">--mode</span> All <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>    <span class="at">--haps</span> <span class="va">${path}</span>/1000G_aDNA_mask_transv_chr1.haps.gz <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>    <span class="at">--sample</span> <span class="va">${path}</span>/1000G_aDNA_mask_transv_chr1.sample.gz <span class="dt">\</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>    <span class="at">--dist</span> <span class="va">${path}</span>/1000G_aDNA_mask_transv_chr1.dist.gz <span class="dt">\</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>    <span class="at">--map</span> <span class="va">${path}</span>/genetic_map_combined_b37_chr1.txt <span class="dt">\</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>    <span class="at">-o</span> 1000G_aDNA_mask_transv_chr1 <span class="dt">\</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>    <span class="at">-m</span> 4e-9 <span class="dt">\</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>    <span class="at">--coal</span> <span class="va">${path}</span>/1000G_auto.coal <span class="dt">\</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>    <span class="at">--sample_ages</span> <span class="va">${path}</span>/sample_ages.txt <span class="dt">\</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>    <span class="at">--memory</span> 5 <span class="dt">\</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>    <span class="at">--seed</span> 1</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-f</span> 1000G_aDNA_mask_transv_chr1.anc</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-f</span> 1000G_aDNA_mask_transv_chr1.mut</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co">#&gt; *********************************************************</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="co">#&gt; Relate</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="co">#&gt;  * Authors: Leo Speidel, Marie Forest, Sinan Shi, Simon Myers.</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="co">#&gt;  * Doc:     https://myersgroup.github.io/relate</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a><span class="co">#&gt; Using:</span></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="co">#&gt;   /opt/homebrew/lib/R/4.3/site-library/twigstats/example//1000G_aDNA_mask_transv_chr1.haps.gz</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a><span class="co">#&gt;   /opt/homebrew/lib/R/4.3/site-library/twigstats/example//1000G_aDNA_mask_transv_chr1.sample.gz</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a><span class="co">#&gt;   /opt/homebrew/lib/R/4.3/site-library/twigstats/example//genetic_map_combined_b37_chr1.txt</span></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a><span class="co">#&gt; with mu = 4e-09 and coal = /opt/homebrew/lib/R/4.3/site-library/twigstats/example//1000G_auto.coal.</span></span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a><span class="co">#&gt; Parsing data..</span></span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a><span class="co">#&gt; Warning: Will use min 0.00019GB of hard disc.</span></span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a><span class="co">#&gt; CPU Time spent: 5.212840s; Max Memory usage: 57Mb.</span></span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a><span class="co">#&gt; Read 90 haplotypes with 229231 SNPs per haplotype.</span></span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a><span class="co">#&gt; Expected minimum memory usage: 1.5Gb.</span></span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a><span class="co">#&gt; Starting chunk 0 of 0.</span></span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-48"><a href="#cb7-48" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-49"><a href="#cb7-49" tabindex="-1"></a><span class="co">#&gt; Painting sequences...</span></span>
<span id="cb7-50"><a href="#cb7-50" tabindex="-1"></a><span class="co">#&gt; CPU Time spent: 6.716679s; Max Memory usage: 93Mb.</span></span>
<span id="cb7-51"><a href="#cb7-51" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-52"><a href="#cb7-52" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-53"><a href="#cb7-53" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-54"><a href="#cb7-54" tabindex="-1"></a><span class="co">#&gt; Estimating topologies of AncesTrees in sections 0-0...</span></span>
<span id="cb7-55"><a href="#cb7-55" tabindex="-1"></a><span class="co">#&gt; [0/0]</span></span>
<span id="cb7-56"><a href="#cb7-56" tabindex="-1"></a><span class="ex">CPU</span> Time spent: 20.474734s<span class="kw">;</span> <span class="ex">Max</span> Memory usage: 1.5e+03Mb.</span>
<span id="cb7-57"><a href="#cb7-57" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-58"><a href="#cb7-58" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-59"><a href="#cb7-59" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-60"><a href="#cb7-60" tabindex="-1"></a><span class="co">#&gt; Propagating mutations across AncesTrees...</span></span>
<span id="cb7-61"><a href="#cb7-61" tabindex="-1"></a><span class="co">#&gt; CPU Time spent: 26.978777s; Max Memory usage: 1.5e+03Mb.</span></span>
<span id="cb7-62"><a href="#cb7-62" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-63"><a href="#cb7-63" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-64"><a href="#cb7-64" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-65"><a href="#cb7-65" tabindex="-1"></a><span class="co">#&gt; Inferring branch lengths of AncesTrees in sections 0-0...</span></span>
<span id="cb7-66"><a href="#cb7-66" tabindex="-1"></a><span class="co">#&gt; [0/0]</span></span>
<span id="cb7-67"><a href="#cb7-67" tabindex="-1"></a><span class="ex">CPU</span> Time spent: 156.776092s<span class="kw">;</span> <span class="ex">Max</span> Memory usage: 1.5e+03Mb.</span>
<span id="cb7-68"><a href="#cb7-68" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-69"><a href="#cb7-69" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-70"><a href="#cb7-70" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-71"><a href="#cb7-71" tabindex="-1"></a><span class="co">#&gt; Combining AncesTrees in Sections...</span></span>
<span id="cb7-72"><a href="#cb7-72" tabindex="-1"></a><span class="co">#&gt; CPU Time spent: 160.431945s; Max Memory usage: 1.5e+03Mb.</span></span>
<span id="cb7-73"><a href="#cb7-73" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-74"><a href="#cb7-74" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-75"><a href="#cb7-75" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-76"><a href="#cb7-76" tabindex="-1"></a><span class="co">#&gt; Finalizing...</span></span>
<span id="cb7-77"><a href="#cb7-77" tabindex="-1"></a><span class="co">#&gt; Number of not mapping SNPs: 414</span></span>
<span id="cb7-78"><a href="#cb7-78" tabindex="-1"></a><span class="co">#&gt; Number of flipped SNPs    : 77</span></span>
<span id="cb7-79"><a href="#cb7-79" tabindex="-1"></a><span class="co">#&gt; CPU Time spent: 164.650103s; Max Memory usage: 1.5e+03Mb.</span></span>
<span id="cb7-80"><a href="#cb7-80" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-81"><a href="#cb7-81" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-82"><a href="#cb7-82" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-83"><a href="#cb7-83" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb7-84"><a href="#cb7-84" tabindex="-1"></a><span class="co">#&gt; ---------------------------------------------------------</span></span>
<span id="cb7-85"><a href="#cb7-85" tabindex="-1"></a><span class="co">#&gt; *********************************************************</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-twigstats">Running Twigstats<a class="anchor" aria-label="anchor" href="#running-twigstats"></a>
</h2>
<p>Finally, once we have inferred Relate trees, we can now run
Twigstats. We will aim to infer an admixture in Anglo-Saxon individuals
between the preceding British Iron/Roman Age and Early Iron Age
Scandinavia.</p>
<p>To illustrate a power gain by using Twigstats, we will run Twigstats
in two different ways: First, we will compute regular f2-statistics and
second, we will compute f2-statistics on only recent coalescences.</p>
<p>We first load the Twigstats package and assign filenames to
variables.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/leospeidel/twigstats" class="external-link">twigstats</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">file_anc</span>  <span class="op">&lt;-</span> <span class="st">"1000G_aDNA_mask_transv_chr1.anc.gz"</span></span>
<span><span class="va">file_mut</span>  <span class="op">&lt;-</span> <span class="st">"1000G_aDNA_mask_transv_chr1.mut.gz"</span></span>
<span><span class="va">poplabels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"example/1000GP_Phase3_sub_aDNA.poplabels"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span></span>
<span><span class="va">file_map</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"example/genetic_map_combined_b37_chr1.txt"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span></span></code></pre></div>
<p>Next, we compute regular f2 statistics without a time cutoff, and
then infer the admixture proportion using an f4-ratio statistic:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Calculate f2s between all pairs of populations</span></span>
<span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_Relate.html">f2_blocks_from_Relate</a></span><span class="op">(</span><span class="va">file_anc</span>, <span class="va">file_mut</span>, <span class="va">poplabels</span>, <span class="va">file_map</span>, use_muts <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"England_Saxon.SG"</span>, popI<span class="op">=</span><span class="st">"Ireland_BA.SG"</span>, pop1<span class="op">=</span><span class="st">"Britain.IronRoman.SG"</span>, pop2<span class="op">=</span><span class="st">"Scandinavian_Peninsula_EIA(I).SG"</span>, popO<span class="op">=</span><span class="st">"YRI"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO          popI                 pop1                             pop2</span></span>
<span><span class="co">#&gt; 1  YRI Ireland_BA.SG Britain.IronRoman.SG Scandinavian_Peninsula_EIA(I).SG</span></span>
<span><span class="co">#&gt;               popX      val       se</span></span>
<span><span class="co">#&gt; 1 England_Saxon.SG 3.083356 63.55952</span></span></code></pre></div>
<p>As we can see, the inferred admixture proportion is not between 0 and
1 and the standard errors are extremely large, giving us no meaningful
inference.</p>
<p>We now compute f2 statistics on lineages of the past 500 generations
using Twigstats:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Use a cutoff of 500 generations</span></span>
<span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/f2_blocks_from_Relate.html">f2_blocks_from_Relate</a></span><span class="op">(</span><span class="va">file_anc</span>, <span class="va">file_mut</span>, <span class="va">poplabels</span>, <span class="va">file_map</span>, t <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"England_Saxon.SG"</span>, popI<span class="op">=</span><span class="st">"Ireland_BA.SG"</span>, pop1<span class="op">=</span><span class="st">"Britain.IronRoman.SG"</span>, pop2<span class="op">=</span><span class="st">"Scandinavian_Peninsula_EIA(I).SG"</span>, popO<span class="op">=</span><span class="st">"YRI"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   popO          popI                 pop1                             pop2</span></span>
<span><span class="co">#&gt; 1  YRI Ireland_BA.SG Britain.IronRoman.SG Scandinavian_Peninsula_EIA(I).SG</span></span>
<span><span class="co">#&gt;               popX       val        se</span></span>
<span><span class="co">#&gt; 1 England_Saxon.SG 0.4189351 0.1824581</span></span></code></pre></div>
<p>The standard errors are much reduced and we get a meaningful
<b>admixture proportion of approximately 38% Scandinavian EIA
ancestry</b>! Note that the standard error is still relatively large
(though excludes 0 and 1 as feasible proportions). The obvious way to
improve power is to use all other chromosomes; here we only used
chromosome 1. Additionally, we could build Relate trees with more
samples to get more accurate constrains on dates.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Leo Speidel, Pontus Skoglund.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
