<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
The output is in a format that is directly accepted by the admixtools R package to calculate 
f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/)."><title>Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees. — f2_blocks_from_RelateAges • twigstats</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees. — f2_blocks_from_RelateAges"><meta property="og:description" content="This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
The output is in a format that is directly accepted by the admixtools R package to calculate 
f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">twigstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/basic-usage.html">Basic usage</a>
    <a class="dropdown-item" href="../articles/real-data-example.html">A small real data example</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/leospeidel/twigstats/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees.</h1>
      <small class="dont-index">Source: <a href="https://github.com/leospeidel/twigstats/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>f2_blocks_from_RelateAges.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
The output is in a format that is directly accepted by the admixtools R package to calculate 
f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span></span>
<span>  <span class="va">pref</span>,</span>
<span>  <span class="va">file_mut</span>,</span>
<span>  blgsize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transitions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxmiss <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fam <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pops <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tmin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  t <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_undated <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minMAF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  apply_corr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  debug_mode <span class="op">=</span> <span class="fl">0L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>pref</dt>
<dd><p>Prefix of PLINK binary files, assuming filenames of form ${pref}.bed, ${pref}.bim, $pref.fam.</p></dd>


<dt>file_mut</dt>
<dd><p>(Optional) Prefix of filenames of mut files, assuming filenames of form ${file_mut}_chr1.mut(.gz). Chromosome names have to be consistent to those in the PLINK files. If no file is specified, all mutations in the PLINK file are used.</p></dd>


<dt>blgsize</dt>
<dd><p>(Optional) SNP block size in Morgan. Default is 0.05 (50 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.</p></dd>


<dt>transitions</dt>
<dd><p>(Optional) Set this to FALSE to exclude transition SNPs</p></dd>


<dt>maxmiss</dt>
<dd><p>(Optional) Discard SNPs which are missing in a fraction of populations higher than $B!F(Bmaxmiss</p></dd>


<dt>fam</dt>
<dd><p>(Optional) 1d-array assigning individuals to populations. Corresponds to the first column in the fam file and is useful if you want to change population assignments.</p></dd>


<dt>pops</dt>
<dd><p>(Optional) Populations for which data should be extracted. Names need to match the first column in the fam file (or fam option below)</p></dd>


<dt>chrs</dt>
<dd><p>(Optional) List chromosome names to use.</p></dd>


<dt>tmin</dt>
<dd><p>(Optional) Minimum time cutof in generations. Any mutations younger than tmin will be excluded from the analysis. Default: t = 0.</p></dd>


<dt>t</dt>
<dd><p>(Optional) Time cutoff in generations. Any mutations older that t will be excluded from the analysis. Default: t = Inf.</p></dd>


<dt>include_undated</dt>
<dd><p>(Optional) Include mutations that are not dated. Default: false.</p></dd>


<dt>minMAF</dt>
<dd><p>(Optional) minimum minor allele count. Default: 1.</p></dd>


<dt>apply_corr</dt>
<dd><p>(Optional) Use small sample size correction. Default: true.</p></dd>


<dt>debug_mode</dt>
<dd><p>(Optional) Prints progress used for debugging.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>3d array of dimension #groups x #groups x #blocks containing f2 statistics. Analogous to output of f2_from_geno in admixtools.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span>,<span class="st">"/"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pref</span> <span class="op">&lt;-</span> <span class="st">"msprime_ad0.8_split250_1"</span></span></span>
<span class="r-in"><span><span class="va">file_plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix</span></span></span>
<span class="r-in"><span><span class="va">file_mut</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix (here same name as plink file but can be different)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Compute f2 statistics between all pairs of populations. You can use pops to only calculate f2s between specified populations.</span></span></span>
<span class="r-in"><span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   popO popI pop1 pop2 popX      val        se</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   P4   P1   P2   P3   PX 1.082681 0.2931867</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Use a cutoff of 500 generations</span></span></span>
<span class="r-in"><span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span>, t <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   popO popI pop1 pop2 popX      val         se</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   P4   P1   P2   P3   PX 0.810077 0.03097378</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Leo Speidel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

