<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees. — f2_blocks_from_RelateAges • twigstats</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees. — f2_blocks_from_RelateAges"><meta property="og:description" content="This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
The output is in a format that is directly accepted by the admixtools R package to calculate
f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/)."><meta property="og:image" content="https://leospeidel.github.io/twigstats/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-P8RP1CGQ1W"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-P8RP1CGQ1W');
</script></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">twigstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/basic-usage.html">Basic usage</a>
    </li>
    <li>
      <a href="../articles/real-data-example.html">A small real data example</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/leospeidel/twigstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees.</h1>
    <small class="dont-index">Source: <a href="https://github.com/leospeidel/twigstats/blob/HEAD/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
    <div class="hidden name"><code>f2_blocks_from_RelateAges.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
The output is in a format that is directly accepted by the admixtools R package to calculate
f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span></span>
<span>  <span class="va">pref</span>,</span>
<span>  <span class="va">file_mut</span>,</span>
<span>  blgsize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transitions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxmiss <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fam <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pops <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tmin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  t <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_undated <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minMAF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  apply_corr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  debug_mode <span class="op">=</span> <span class="fl">0L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-pref">pref<a class="anchor" aria-label="anchor" href="#arg-pref"></a></dt>
<dd><p>Prefix of PLINK binary files, assuming filenames of form ${pref}.bed, ${pref}.bim, ${pref}.fam.</p></dd>


<dt id="arg-file-mut">file_mut<a class="anchor" aria-label="anchor" href="#arg-file-mut"></a></dt>
<dd><p>(Optional) Prefix of filenames of mut files, assuming filenames of form ${file_mut}_chr1.mut(.gz). Chromosome names have to be consistent to those in the PLINK files. If no file is specified, all mutations in the PLINK file are used.</p></dd>


<dt id="arg-blgsize">blgsize<a class="anchor" aria-label="anchor" href="#arg-blgsize"></a></dt>
<dd><p>(Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.</p></dd>


<dt id="arg-transitions">transitions<a class="anchor" aria-label="anchor" href="#arg-transitions"></a></dt>
<dd><p>(Optional) Set this to FALSE to exclude transition SNPs</p></dd>


<dt id="arg-maxmiss">maxmiss<a class="anchor" aria-label="anchor" href="#arg-maxmiss"></a></dt>
<dd><p>(Optional) Discard SNPs which are missing in a fraction of populations higher than maxmiss</p></dd>


<dt id="arg-fam">fam<a class="anchor" aria-label="anchor" href="#arg-fam"></a></dt>
<dd><p>(Optional) 1d-array assigning individuals to populations. Corresponds to the first column in the fam file and is useful if you want to change population assignments.</p></dd>


<dt id="arg-pops">pops<a class="anchor" aria-label="anchor" href="#arg-pops"></a></dt>
<dd><p>(Optional) Populations for which data should be extracted. Names need to match the first column in the fam file (or fam option below)</p></dd>


<dt id="arg-chrs">chrs<a class="anchor" aria-label="anchor" href="#arg-chrs"></a></dt>
<dd><p>(Optional) List chromosome names to use.</p></dd>


<dt id="arg-tmin">tmin<a class="anchor" aria-label="anchor" href="#arg-tmin"></a></dt>
<dd><p>(Optional) Minimum time cutof in generations. Any mutations younger than tmin will be excluded from the analysis. Default: t = 0.</p></dd>


<dt id="arg-t">t<a class="anchor" aria-label="anchor" href="#arg-t"></a></dt>
<dd><p>(Optional) Time cutoff in generations. Any mutations older that t will be excluded from the analysis. Default: t = Inf.</p></dd>


<dt id="arg-include-undated">include_undated<a class="anchor" aria-label="anchor" href="#arg-include-undated"></a></dt>
<dd><p>(Optional) Include mutations that are not dated. Default: false.</p></dd>


<dt id="arg-minmaf">minMAF<a class="anchor" aria-label="anchor" href="#arg-minmaf"></a></dt>
<dd><p>(Optional) minimum minor allele count. Default: 1.</p></dd>


<dt id="arg-apply-corr">apply_corr<a class="anchor" aria-label="anchor" href="#arg-apply-corr"></a></dt>
<dd><p>(Optional) Use small sample size correction. Default: true.</p></dd>


<dt id="arg-debug-mode">debug_mode<a class="anchor" aria-label="anchor" href="#arg-debug-mode"></a></dt>
<dd><p>(Optional) Prints progress used for debugging.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>3d array of dimension #groups x #groups x #blocks containing f2 statistics. Analogous to output of f2_from_geno in admixtools.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"sim/"</span>, package <span class="op">=</span> <span class="st">"twigstats"</span><span class="op">)</span>,<span class="st">"/"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pref</span> <span class="op">&lt;-</span> <span class="st">"msprime_ad0.8_split250_1"</span></span></span>
<span class="r-in"><span><span class="va">file_plink</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix</span></span></span>
<span class="r-in"><span><span class="va">file_mut</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">path</span>,<span class="va">pref</span><span class="op">)</span> <span class="co">#only need prefix (here same name as plink file but can be different)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gunzip "</span>, <span class="va">file_plink</span>, <span class="st">".bim.gz"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Compute f2 statistics between all pairs of populations. You can use pops to only calculate f2s between specified populations.</span></span></span>
<span class="r-in"><span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   popO popI pop1 pop2 popX      val        se</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   P4   P1   P2   P3   PX 1.082681 0.2931867</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Use a cutoff of 500 generations</span></span></span>
<span class="r-in"><span><span class="va">f2_blocks</span> <span class="op">&lt;-</span> <span class="fu">f2_blocks_from_RelateAges</span><span class="op">(</span>pref <span class="op">=</span> <span class="va">file_plink</span>, <span class="va">file_mut</span>, t <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="f-statistics.html">f4_ratio</a></span><span class="op">(</span><span class="va">f2_blocks</span>, popX<span class="op">=</span><span class="st">"PX"</span>, popI<span class="op">=</span><span class="st">"P1"</span>, pop1<span class="op">=</span><span class="st">"P2"</span>, pop2<span class="op">=</span><span class="st">"P3"</span>, popO<span class="op">=</span><span class="st">"P4"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   popO popI pop1 pop2 popX      val         se</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   P4   P1   P2   P3   PX 0.810077 0.03097378</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leo Speidel, Pontus Skoglund.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

