# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Function to calculate D statistics from Relate trees.
#'
#' This function will calculate D statistics in blocks of prespecified size for all pairs of populations specified in the poplabels file.
#' Please refer to the Relate documentation for input file formats (https://myersgroup.github.io/relate/).
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param poplabels Filename of poplabels file
#' @param pop1 Population 1
#' @param pop2 Population 2
#' @param pop3 Population 3
#' @param pop4 Population 4
#' @param chrs (Optional) Vector of chromosome IDs
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance. If blgsize is negative, every tree is its own block.
#' @param file_map (Optional) File prefix of recombination map. Not needed if blgsize is given in base-pairs, i.e. blgsize > 100
#' @param mu (Optional) Per base per generation mutation rate to scale f2 values. Default: 1.25e-8
#' @param t (Optional) Time cutoff in generations. Default: Inf
#' @param tmin (Optional) Minimum time cutof in generations. Any lineages younger than tmin will be excluded from the analysis. Default: t = 0.
#' @param minMAF (Optional) Minimum frequency cutoff. Default: 1 (i.e. excl singletons)
#' @param use_muts (Optional) Calculate traditional f2 statistics by only using mutations mapped to Relate trees. Default: false.
#' @param transitions (Optional) Set this to FALSE to exclude transition SNPs. Only meaningful with use_muts
#' @return Data frame accepted by jackknife function.
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #Calculate D
#' D_blocks <- D_from_Relate(file_anc, file_mut, poplabels, "P4", "P3", "P1", "P2", file_map)
#' print(jackknife(D_blocks))
#' D_blocks <- D_from_Relate(file_anc, file_mut, poplabels, "P4", "P3", "P1", "PX", file_map)
#' print(jackknife(D_blocks))
#' 
#' #Calculate D with time cutoff
#' D_blocks <- D_from_Relate(file_anc, file_mut, poplabels, "P4", "P3", "P1", "PX", file_map, t = 1000)
#' print(jackknife(D_blocks))
#'
#' #Calculate D with mutations only (and time cutoff)
#' D_blocks <- D_from_Relate(file_anc, file_mut, poplabels, "P4", "P3", "P1", "PX", file_map, use_muts = T, t = 1000)
#' print(jackknife(D_blocks))
#' @export
D_from_Relate <- function(file_anc, file_mut, poplabels, pop1, pop2, pop3, pop4, file_map = NULL, chrs = NULL, blgsize = NULL, mu = NULL, tmin = NULL, t = NULL, transitions = NULL, use_muts = NULL, minMAF = NULL) {
    .Call(`_twigstats_D_from_Relate`, file_anc, file_mut, poplabels, pop1, pop2, pop3, pop4, file_map, chrs, blgsize, mu, tmin, t, transitions, use_muts, minMAF)
}

#' Function to calculate f2 statistics from Relate trees for pairs of populations specified in poplabels.
#'
#' This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the poplabels file.
#' Please refer to the Relate documentation for input file formats (https://myersgroup.github.io/relate/).
#' The output is in a format that is directly accepted by the admixtools R package to calculate 
#' f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/).
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param poplabels Filename of poplabels file
#' @param chrs (Optional) Vector of chromosome IDs
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance. If blgsize is negative, every tree is its own block.
#' @param file_map (Optional) File prefix of recombination map. Not needed if blgsize is given in base-pairs, i.e. blgsize > 100
#' @param mu (Optional) Per base per generation mutation rate to scale f2 values. Default: 1.25e-8
#' @param t (Optional) Time cutoff in generations. Default: Inf
#' @param tmin (Optional) Minimum time cutof in generations. Any lineages younger than tmin will be excluded from the analysis. Default: t = 0.
#' @param minMAF (Optional) Minimum frequency cutoff. Default: 1 (i.e. excl singletons)
#' @param use_muts (Optional) Calculate traditional f2 statistics by only using mutations mapped to Relate trees. Default: false.
#' @param transitions (Optional) Set this to FALSE to exclude transition SNPs. Only meaningful with use_muts
#' @param apply_corr (Optional) Use small sample size correction. Default: true.
#' @param dump_blockpos (Optional) Filename of blockpos file.
#' @return 3d array of dimension #groups x #groups x #blocks. Analogous to output of f2_from_geno in admixtools.
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #Calculate f2s between all pairs of populations
#' f2_blocks <- f2_blocks_from_Relate(file_anc, file_mut, poplabels, file_map)
#' f4_ratio(f2_blocks, popX="PX", popI="P1", pop1="P2", pop2="P3", popO="P4")
#'
#' #Use a cutoff of 500 generations
#' f2_blocks <- f2_blocks_from_Relate(file_anc, file_mut, poplabels, file_map, t = 500)
#' f4_ratio(f2_blocks, popX="PX", popI="P1", pop1="P2", pop2="P3", popO="P4")
#' @export
f2_blocks_from_Relate <- function(file_anc, file_mut, poplabels, file_map = NULL, chrs = NULL, blgsize = NULL, mu = NULL, tmin = NULL, t = NULL, transitions = NULL, use_muts = NULL, minMAF = NULL, dump_blockpos = NULL, apply_corr = NULL) {
    .Call(`_twigstats_f2_blocks_from_Relate`, file_anc, file_mut, poplabels, file_map, chrs, blgsize, mu, tmin, t, transitions, use_muts, minMAF, dump_blockpos, apply_corr)
}

#' Function to calculate f2 statistics from plink files, ascertained using mutation ages in Relate trees.
#'
#' This function will calculate f2 statistics in blocks of prespecified size for all pairs of populations specified in the input files.
#' Input is assumed to be in PLINK binary format https://www.cog-genomics.org/plink/1.9/formats#bed and mutation ages are in Relate mut format https://myersgroup.github.io/relate/.
#' The output is in a format that is directly accepted by the admixtools R package to calculate 
#' f2, f3, f4, f4ratio, D statistics and more (https://uqrmaie1.github.io/admixtools/).
#'
#' @param pref Prefix of PLINK binary files, assuming filenames of form $\{pref\}.bed, $\{pref\}.bim, $\{pref\}.fam.
#' @param file_mut (Optional) Prefix of filenames of mut files, assuming filenames of form $\{file_mut\}_chr1.mut(.gz). Chromosome names have to be consistent to those in the PLINK files. If no file is specified, all mutations in the PLINK file are used.
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.
#' @param t (Optional) Time cutoff in generations. Any mutations older that t will be excluded from the analysis. Default: t = Inf.
#' @param tmin (Optional) Minimum time cutof in generations. Any mutations younger than tmin will be excluded from the analysis. Default: t = 0.
#' @param transitions (Optional) Set this to FALSE to exclude transition SNPs
#' @param maxmiss (Optional) Discard SNPs which are missing in a fraction of populations higher than maxmiss
#' @param pops (Optional) Populations for which data should be extracted. Names need to match the first column in the fam file (or fam option below)
#' @param fam (Optional) 1d-array assigning individuals to populations. Corresponds to the first column in the fam file and is useful if you want to change population assignments.
#' @param chrs (Optional) List chromosome names to use.
#' @param include_undated (Optional) Include mutations that are not dated. Default: false.
#' @param minMAF (Optional) minimum minor allele count. Default: 1.
#' @param apply_corr (Optional) Use small sample size correction. Default: true.
#' @param debug_mode (Optional) Prints progress used for debugging.
#' @return 3d array of dimension #groups x #groups x #blocks containing f2 statistics. Analogous to output of f2_from_geno in admixtools.
#' @examples
#' path <- paste0(system.file("sim/", package = "twigstats"),"/")
#' pref <- "msprime_ad0.8_split250_1"
#' file_plink <- paste0(path,pref) #only need prefix
#' file_mut  <- paste0(path,pref) #only need prefix (here same name as plink file but can be different)
#'
#' system(paste0("gunzip ", file_plink, ".bim.gz"))
#'
#' #Compute f2 statistics between all pairs of populations. You can use pops to only calculate f2s between specified populations.
#' f2_blocks <- f2_blocks_from_RelateAges(pref = file_plink, file_mut)
#' f4_ratio(f2_blocks, popX="PX", popI="P1", pop1="P2", pop2="P3", popO="P4")
#'
#' #Use a cutoff of 500 generations
#' f2_blocks <- f2_blocks_from_RelateAges(pref = file_plink, file_mut, t = 500)
#' f4_ratio(f2_blocks, popX="PX", popI="P1", pop1="P2", pop2="P3", popO="P4")
#' @export
f2_blocks_from_RelateAges <- function(pref, file_mut, blgsize = NULL, transitions = NULL, maxmiss = NULL, fam = NULL, pops = NULL, chrs = NULL, tmin = NULL, t = NULL, include_undated = NULL, minMAF = NULL, apply_corr = NULL, debug_mode = 0L) {
    .Call(`_twigstats_f2_blocks_from_RelateAges`, pref, file_mut, blgsize, transitions, maxmiss, fam, pops, chrs, tmin, t, include_undated, minMAF, apply_corr, debug_mode)
}

#' Function to calculate Fst from Relate trees for pairs of populations specified in poplabels.
#'
#' This function will calculate Fst in blocks of prespecified size for all pairs of populations specified in the poplabels file.
#' Please refer to the Relate documentation for input file formats (https://myersgroup.github.io/relate/).
#' The output is in the same format as for f2_blocks_from_Relate. 
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param poplabels Filename of poplabels file
#' @param chrs (Optional) Vector of chromosome IDs
#' @param blgsize (Optional) SNP block size. Default is 10000 bases. If blgsize is <100 then it will be interpreted in Morgan. If blgsize is 100 or greater, if will be interpreted as base pair distance. If blgsize is negative, every tree is its own block.
#' @param file_map (Optional) File prefix of recombination map. Not needed if blgsize is given in base-pairs, i.e. blgsize > 100
#' @param mu (Optional) Per base per generation mutation rate to scale f2 values. Default: 1.25e-8
#' @param t (Optional) Time cutoff in generations. Default: Inf
#' @param tmin (Optional) Minimum time cutof in generations. Any lineages younger than tmin will be excluded from the analysis. Default: t = 0.
#' @param minMAF (Optional) Minimum frequency cutoff. Default: 1 (i.e. excl singletons)
#' @param use_muts (Optional) Calculate traditional f2 statistics by only using mutations mapped to Relate trees. Default: false.
#' @param transitions (Optional) Set this to FALSE to exclude transition SNPs. Only meaningful with use_muts
#' @param apply_corr (Optional) Use small sample size correction. Default: true.
#' @param dump_blockpos (Optional) Filename of blockpos file.
#' @return 3d array of dimension #groups x #groups x #blocks. Analogous to output of f2_from_geno in admixtools.
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #Calculate f2s between all pairs of populations
#' Fst_blocks <- Fst_blocks_from_Relate(file_anc, file_mut, poplabels, file_map)
#'
#' #Use a cutoff of 500 generations
#' Fst_blocks <- Fst_blocks_from_Relate(file_anc, file_mut, poplabels, file_map, dump_blockpos = "test.pos", t = 500)
#' @export
Fst_blocks_from_Relate <- function(file_anc, file_mut, poplabels, file_map = NULL, chrs = NULL, blgsize = NULL, mu = NULL, tmin = NULL, t = NULL, transitions = NULL, use_muts = NULL, minMAF = NULL, Fst = NULL, dump_blockpos = NULL, apply_corr = NULL) {
    .Call(`_twigstats_Fst_blocks_from_Relate`, file_anc, file_mut, poplabels, file_map, chrs, blgsize, mu, tmin, t, transitions, use_muts, minMAF, Fst, dump_blockpos, apply_corr)
}

#' Function to calculates mean TMRCAs from Relate trees for pairs of populations specified in poplabels.
#'
#' This function will calculate TMRCAs in blocks of prespecified size for all pairs of populations specified in the poplabels file.
#' Please refer to the Relate documentation for input file formats (https://myersgroup.github.io/relate/).
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_out Filename of output file.
#' @param poplabels Filename of poplabels file
#' @param t (Optional) Time cutoff in generations. Any coalescences older that t will be set to t in the analysis. Default: t = Inf.
#' @param chrs (Optional) Vector of chromosome IDs
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.
#' @param file_map (Optional) File prefix of recombination map. Not needed if blgsize is given in base-pairs, i.e. blgsize > 100
#' @return 3d array of dimension #groups x #groups x #blocks. Analogous to output of f2_from_geno in admixtools.
#' @keywords internal
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #Calculate f2s between all pairs of populations
#' TMRCA_from_Relate(file_anc, file_mut, poplabels, file_out = "test", file_map)
#' @export
TMRCA_from_Relate <- function(file_anc, file_mut, poplabels, file_out, file_map = NULL, chrs = NULL, t = NULL, blgsize = NULL) {
    invisible(.Call(`_twigstats_TMRCA_from_Relate`, file_anc, file_mut, poplabels, file_out, file_map, chrs, t, blgsize))
}

#' Function to calculates TMRCA distributions from Relate trees for pairs of populations specified in poplabels.
#'
#' This function will calculate TMRCAs in blocks of prespecified size for all pairs of populations specified in the poplabels file.
#' Please refer to the Relate documentation for input file formats (https://myersgroup.github.io/relate/).
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_out Filename of output file.
#' @param poplabels Filename of poplabels file
#' @param epochs Vector of epoch boundaries. Should start at 0.
#' @param chrs (Optional) Vector of chromosome IDs
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 100 or greater, if will be interpreted as base pair distance rather than centimorgan distance.
#' @param file_map (Optional) File prefix of recombination map. Not needed if blgsize is given in base-pairs, i.e. blgsize > 100
#' @return 3d array of dimension #groups x #groups x #blocks. Analogous to output of f2_from_geno in admixtools.
#' @keywords internal
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #Calculate f2s between all pairs of populations
#' TMRCAdist_from_Relate(file_anc, file_mut, poplabels, file_out = "test", file_map, epochs = c(0,10^seq(3,7,length.out=49)/28))
#' @export
TMRCAdist_from_Relate <- function(file_anc, file_mut, poplabels, file_out, epochs, file_map = NULL, chrs = NULL, t = NULL, blgsize = NULL) {
    invisible(.Call(`_twigstats_TMRCAdist_from_Relate`, file_anc, file_mut, poplabels, file_out, epochs, file_map, chrs, t, blgsize))
}

#' Chromosome painting using genealogies.
#'
#' This function outputs the first coalescence with an individual from a pre-specified group identity along the genome.
#' If the first such coalescnece involves several copying candidates, a random haplotype is chosen.
#' Output is in GLOBEtrotter format.
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_map File prefix of recombination map.
#' @param file_out File prefix of output files
#' @param poplabels Filename of poplabels file
#' @param blgsize (Optional) SNP block size in Morgan. Default is 0.05 (5 cM). If blgsize is 1 or greater, if will be interpreted as base pair distance rather than centimorgan distance.
#' @param pops (Optional) Populations for which data should be extracted. Names need to match the second column of the poplabels file
#' @param chrs (Optional) Vector of chromosome IDs
#' @return void. Write three files idfile, paint, rec to disc. 
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#' file_map  <- system.file("sim/genetic_map_combined_b37_chr1.txt.gz", package = "twigstats")
#'
#' #define populations to paint against:
#' pops <- c("P1","P2","P3","P4")
#'
#' Painting(file_anc, file_mut, file_map, file_out = "test", poplabels, blgsize = 1e-5)
#' @export
Painting <- function(file_anc, file_mut, file_map, file_out, poplabels, blgsize = NULL, pops = NULL, chrs = NULL) {
    invisible(.Call(`_twigstats_Painting`, file_anc, file_mut, file_map, file_out, poplabels, blgsize, pops, chrs))
}

#' Function to compute genome-wide copying proportions.
#'
#' This function takes the output of the function Painting and computes the genome-wide 'copying vectors',
#' i.e., the proportion of the genome copied from each reference population.
#'
#' @param filename_painting A vector containing filenames of painting profiles. These are the outputs of Painting.
#' @param filename_idfile The filename of the idfile, which is an output of Painting.
#' @param nboot Number of bootstrap samples to generate.
#' @param blocksize Number of blocks to combine for the bootstrap. For example, if Painting was run with a blgsize of 1e-5 Morgans, blocksize should be 5000 to achieve a blocksize of 5cM.
#' @param use_IDs If TRUE, compute profiles for each sample. If FALSE (default), compute profiles for each group as specified in the second column of the idfile.
#' @return A DataFrame with the copying proportions per bootstrap sample.
#' @examples
#' # Example usage in R:
#' path      <- paste0(system.file("test/", package = "twigstats"),"/")
#' prefix    <- "test" # prefix of files under path
#' df <- PaintingProfile(c(paste0(path, prefix, "_painting.txt.gz")),
#'                          paste0(path, prefix, "_idfile.txt.gz"),
#'                          nboot = 10, blocksize = 5000)
#' head(df)
#' @export
PaintingProfile <- function(filename_painting, filename_idfile, nboot, blocksize, use_IDs = FALSE) {
    .Call(`_twigstats_PaintingProfile`, filename_painting, filename_idfile, nboot, blocksize, use_IDs)
}

#' Function to calculate first coalescence copying vector for each population specified in poplabels file.
#'
#' This function calculates the proportion of the genome where the first coalescence is with group x.
#' It returns a version where coalescence with own group is counted, and one where the own group is excluded.
#'
#' @param file_anc Filename of anc file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param file_mut Filename of mut file. If chrs is specified, this should only be the prefix, resulting in filenames of $\{file_anc\}_chr$\{chr\}.anc(.gz).
#' @param poplabels Filename of poplabels file
#' @param pops (Optional) Populations for which data should be extracted. Names need to match the second column of the poplabels file
#' @param chrs (Optional) Vector of chromosome IDs
#' @return 2d array of dimension #groups x #groups.
#' @keywords internal
#' @examples
#' file_anc  <- system.file("sim/msprime_ad0.8_split250_1_chr1.anc.gz", package = "twigstats")
#' file_mut  <- system.file("sim/msprime_ad0.8_split250_1_chr1.mut.gz", package = "twigstats")
#' poplabels <- system.file("sim/msprime_ad0.8_split250_1.poplabels", package = "twigstats")
#'
#' #Calculate f2s between all pairs of populations
#' nn_res <- ExpPaintingProfile(file_anc, file_mut, poplabels)
#' if(require(lsei)){
#'   pnnls(t(nn_res[c("P2","P3"),c("P1","P2","P3","P4"),1]),(nn_res["PX",c("P1","P2","P3","P4"),1]), sum = 1)$x
#' }
#'
#' @export
ExpPaintingProfile <- function(file_anc, file_mut, poplabels, pops = NULL, chrs = NULL) {
    .Call(`_twigstats_ExpPaintingProfile`, file_anc, file_mut, poplabels, pops, chrs)
}

